### ISSUE TITLE
DHCP client supposed to add iface during the "rebinding and renewing" state

### ISSUE THREAD

**Issue by RamPingle-eaton** [zephyrproject-rtos/zephyr#75926]
Dear Team,

I am currently engaged in developing an IoT application using Zephyr version 3.6.0. I have noticed that when the Ethernet cable is attached and subsequently detached, my application ceases to function properly. It consistently fails in DNS resolution.

**<dbg> net_dns_resolve: dns_resolve_name: (XYZ Task): DNS id will be 19271
<dbg> net_dns_resolve: dns_write: (XYZ Task): [0] submitting work to server idx 0 for id 19271 hash 907
<dbg> net_dns_resolve: dns_write: (XYZ Task): Cannot send query (-22)
<dbg> net_dns_resolve: dns_resolve_name: (XYZ Task): DNS query failed 1 times
<dbg> BLR_MQTT: Get_Mqtt_Broker_Addrinfo: DNS not resolved for XYZ.net (returned -11), retrying
<dbg> net_dns_resolve: dns_resolve_name: (XYZ Task): DNS id will be 62627
<dbg> net_dns_resolve: dns_write: (XYZ Task): [0] submitting work to server idx 0 for id 62627 hash 907
<dbg> net_dns_resolve: dns_write: (XYZ Task): Cannot send query (-22)
<dbg> net_dns_resolve: dns_resolve_name: (XYZ Task): DNS query failed 1 times
<dbg> BLR_MQTT: Get_Mqtt_Broker_Addrinfo: DNS not resolved for XYZ.net (returned -11), retrying
<dbg> net_dns_resolve: dns_resolve_name: (XYZ Task): DNS id will be 29578
<dbg> net_dns_resolve: dns_write: (XYZ Task): [0] submitting work to server idx 0 for id 29578 hash 907
<dbg> net_dns_resolve: dns_write: (XYZ Task): Cannot send query (-22)
<dbg> net_dns_resolve: dns_resolve_name: (XYZ Task): DNS query failed 1 times**

Upon investigating the code, I discovered that the DHCP client removes the “iface” when it receives the “**NET_EVENT_IF_DOWN**” event. When it receives the “**NET_EVENT_IF_UP**” event, it sends a DHCP request and enters into a rebinding/renewing state. In the function “**dhcpv4_handle_msg_ack**”, the DHCP module only adds the “iface” back to the system when it is in the “NET_DHCPV4_REQUESTING” state. However, the current state is “NET_DHCPV4_RENEWING”, so it does not add the “iface” back.

To address this issue, I have modified the code as follows:

```
static void dhcpv4_handle_msg_ack(struct net_if *iface)
{
	switch (iface->config.dhcpv4.state) {
	case NET_DHCPV4_DISABLED:
	case NET_DHCPV4_INIT:
	case NET_DHCPV4_SELECTING:
	case NET_DHCPV4_BOUND:
		break;
	case NET_DHCPV4_REQUESTING:
	case NET_DHCPV4_RENEWING:
	case NET_DHCPV4_REBINDING:
		NET_INFO("Received: %s",
			 net_sprint_ipv4_addr(&iface->config.dhcpv4.requested_ip));

		if (!net_if_ipv4_addr_add(iface,
					  &iface->config.dhcpv4.requested_ip,
					  NET_ADDR_DHCP,
					  iface->config.dhcpv4.lease_time)) {
			NET_DBG("Failed to add IPv4 addr to iface %p", iface);
			return;
		}
		dhcpv4_enter_bound(iface);
		break;
	}
}

```
This modification appears to resolve the issue effectively. However, I would appreciate it if you could review this and let me know if there is anything that I might have overlooked.

**Comment by rlubos** [comment]
Duplicate of https://github.com/zephyrproject-rtos/zephyr/issues/75789? The fix has been backported to v3.6-branch

**Comment by carlescufi** [comment]
Duplicate of https://github.com/zephyrproject-rtos/zephyr/issues/75789
